### PR TITLE
Added lua functions and Fixed bug(?)

### PR DESCRIPTION
**Pull request by SeitaHigashi**
I added the lua function.

I found the bug when coding. So, I fixed it, but this updating may be not enough.
If you use the expr option when mapping, we can see the bug.

### CODE REVIEW

**Comment by asiryk** [comment]
Hi, @SeitaHigashi again! Thank you for adding the Lua API, I was thinking to add it too

File: lua/auto-hlsearch.lua [hunk 1]
```diff
@@ -1,23 +1,39 @@
...
 }
 
--- Remap provided keys in order to use :AutoHlsearch command
+-- Remap provided keys in order to use activate() function
 -- while keeping the user's keymap configuration
 local function remap_keys(keys)
   local function set(lhs, keymap)
-    local cmd = ":AutoHlsearch<CR>"
-    if keymap and keymap.rhs then
+    if keymap then
```
Comment (asiryk) [inline lua/auto-hlsearch.lua:11]: there was a bug indeed. the remap_keys would override user's keymap if they have lua function there

File: lua/auto-hlsearch.lua [hunk 2]
```diff
@@ -1,23 +1,39 @@
...
-      vim.api.nvim_set_keymap("n", lhs, string.format("%s%s", cmd, keymap.rhs), opts)
+      -- We need to consider the remmaping when use expr options
+      if keymap.expr == 1 then
+        -- For vimscript function
+        if keymap.rhs then
+          vim.keymap.set("n", lhs, function () M.activate() return vim.api.nvim_eval(keymap.rhs) end , opts)
+
+        -- For lua function
+        elseif keymap.callback then
+          vim.keymap.set("n", lhs, keymap.callback, opts)
```
Comment (asiryk) [inline lua/auto-hlsearch.lua]: there should be M.activate()
```suggestion
          vim.keymap.set("n", lhs, function()
            M.activate()
            keymap.callback()
          end, opts)
```

File: lua/auto-hlsearch.lua [hunk 3]
```diff
@@ -1,23 +1,39 @@
...
-        expr = keymap.expr,
+        expr = true,
         noremap = keymap.noremap,
         nowait = keymap.nowait,
         script = keymap.script,
         silent = keymap.silent,
       }
-      vim.api.nvim_set_keymap("n", lhs, string.format("%s%s", cmd, keymap.rhs), opts)
+      -- We need to consider the remmaping when use expr options
+      if keymap.expr == 1 then
```
Comment (asiryk) [inline lua/auto-hlsearch.lua]: why do we need to check it?
if the user has a lua function in a keymap, it doesn't matter if `keymap.expr == 1` or not, the function should be executed anyway. am I right?

File: lua/auto-hlsearch.lua (hunk 2 above)
Comment (SeitaHigashi) [reply lua/auto-hlsearch.lua]: Fixed it. Thank you.

File: lua/auto-hlsearch.lua (hunk 3 above)
Comment (SeitaHigashi) [reply lua/auto-hlsearch.lua]: I see. I have refactored it.

File: lua/auto-hlsearch.lua [hunk 4]
```diff
@@ -1,23 +1,38 @@
...
+      if keymap.callback then
+        vim.keymap.set("n", lhs, function () M.activate() return keymap.callback() end, opts)
+
+      elseif keymap.expr == 1 and keymap.rhs then
+        -- For vimscript function
+        vim.keymap.set("n", lhs, function () M.activate() return vim.api.nvim_eval(keymap.rhs) end , opts)
+
+      -- For vimscript function, not use expr options
+      elseif keymap.rhs then
+          vim.keymap.set("n", lhs, function () M.activate() return keymap.rhs end, opts)
```
Comment (asiryk) [inline lua/auto-hlsearch.lua:30]: Interesting.. if `expr=true`, the return from lua callback would get evaluated? Didn't know about that. I tested it and it works, but is that documented somewhere in the nvim help? I didn't find anything about it.

**Review by asiryk — APPROVED** [review]

File: lua/auto-hlsearch.lua (hunk 3 above)
Comment (asiryk) [reply lua/auto-hlsearch.lua]: Nice, looks much better!

File: lua/auto-hlsearch.lua (hunk 4 above)
Comment (SeitaHigashi) [reply lua/auto-hlsearch.lua:30]: vim.keymap.set()'s help is writen like "Accepts options accepted by the {opts} parameter in nvim_set_keymap()"
then nvim_set_keymap()'s help is writen like the below
When called from Lua, also accepts a "callback" key that takes a Lua function to call when the mapping is executed.
When "expr" is true, "replace_keycodes" (boolean) can be used to replace keycodes in the resulting string (see nvim_replace_termcodes()), and a Lua callback returning `nil` is equivalent to returning an empty string.

Comment (asiryk) [reply lua/auto-hlsearch.lua:30]: Oh, I see, thank you!

**Pull request merged**